shared/src/main/scala/mlscript/pretyper/PreTyper.scala

```diff
@@ -0,0 +1,345 @@
+        val parameters = defn.params.map(extractParameters).getOrElse(Nil)
+        val thisSymbol = new LocalTermSymbol(Var("this"))
+        traverseTypeDefinition(TypeSymbol(defn), defn)(completeScope ++ (thisSymbol :: parameters))
+      }
+      println(thingsToTraverse.iterator.map {
+        case (L(term), _) => term.showDbg
+        case (R(symbol), _) => symbol.name
+      }.mkString("to be traversed: {", ", ", "}"))
+      // Pass 3: Traverse terms collected from the last pass.
+      println("Pass 3")
+      thingsToTraverse.foreach {
+        case (L(term), scope) =>
+          println("traverseTerm: " + term.showDbg)
+          println("scope: " + scope.showLocalSymbols)
+          traverseTerm(term)(scope)
+        case (R(symbol), scope) => symbol.body match {
+          case L(term) =>
+            if (symbol.isFunction) {
+              traverseTerm(term)(completeScope)
+            } else {
+              traverseTerm(term)(scope)
+            }
+          case R(_) => ()
+        }
+      }
+      completeScope
+    }({ scope => s"traverseStatements ==> Scope {${scope.showLocalSymbols}}" })
+
+  def apply(typingUnit: TypingUnit, scope: Scope, name: Str): Scope =
+    trace(s"PreTyper <== $name: ${typingUnit.describe}") {
+      traverseStatements(typingUnit.entities, name, scope)
+    }({ scope => s"PreTyper ==> ${scope.showLocalSymbols}" })
+  
+  /**
+    * Extract types in class signatures. For example, for this piece of code
+    * ```mls
+    * abstract class Option[A]: Some[A] | None
+    * ```
+    * this function returns, `Some` and `None`.
+    *
+    * @param ty a type obtained from `NuTypeDef.sig`
+    * @return a list of type names, without any p
+    */
+  private def extractSignatureTypes(ty: Type): Ls[TypeName] = {
+    @tailrec
+    def rec(acc: Ls[TypeName], ty: Type): Ls[TypeName] = ty match {
+      case tn: TypeName => tn :: acc
+      case AppliedType(tn: TypeName, _) => tn :: acc
+      case Union(lhs, tn: TypeName) => rec(tn :: acc, lhs)
+      case Union(lhs, AppliedType(tn: TypeName, _)) => rec(tn :: acc, lhs)
+      case other =>
+        // Let's not raise warning for now.
+        // raiseWarning(msg"unknown type in signature" -> other.toLoc)
+        Nil
+    }
+    rec(Nil, ty).reverse
+  }
+
+  def extractSuperTypes(parents: Ls[Term]): Ls[Var] = {
+    @tailrec
+    def rec(acc: Ls[Var], rest: Ls[Term]): Ls[Var] =
+      rest match {
+        case Nil => acc.reverse
+        case (nme: Var) :: tail => rec(nme :: acc, tail)
+        case (TyApp(ty, _)) :: tail => rec(acc, ty :: tail)
+        case (App(term, Tup(_))) :: tail => rec(acc, term :: tail)
+        case head :: tail =>
+          raiseWarning(msg"unknown type in parent types: ${head.showDbg}" -> head.toLoc)
+          rec(acc, tail)
+      }
+    rec(Nil, parents)
+  }
+}
+
+object PreTyper {
+  def transitiveClosure[A](graph: Map[A, List[A]])(implicit ord: Ordering[A]): SortedMap[A, List[A]] = {
```

Comment:
Apparently this needs to be removed?

Reply:
No. I realized that the function computes base classes for each class so that we can know two pattern matching branches are overlapped or not.

Reply:
They've already been removed.